engineapi/proxy.go

```diff
@@ -1,168 +1,421 @@
+	client, exist := h.getEngineClientProxy(instanceManagerImage, clientUUID)
+	if !exist {
+		h.logger.Debugf("Engine client proxy(%v) for %v is already gone", clientUUID, instanceManagerImage)
+		return
+	}
+
+	log := client.getLogger()
+
+	nextGC := client.getNextGC()
+	now := time.Now()
+	if nextGC == nil || now.Before(*nextGC) {
+		return
+	}
+
+	msg := ""
+	if client.getInCall() != 0 {
+		msg = fmt.Sprintf("Cannot stop engine client proxy with %v still in call", client.getInCall())
+		if client.getGCRetry() < ProxyGCMaxRetry {
+			log.Warnf(msg)
+			client.setNextGC()
+			return
+		}
+	}
+
+	log.Debug("Stopping engine client proxy")
+	if err := client.stop(); err != nil {
+		msg = "Failed to stop engine client proxy"
+		if client.getGCRetry() < ProxyGCMaxRetry {
+			log.WithError(err).Warnf(msg)
+			client.setNextGC()
+			return
+		}
+	}
+
+	if msg != "" {
+		log.Warn("Exceed max retry limit, force delete engine client proxy")
+	}
+
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	delete(h.proxyPools[instanceManagerImage].clients, clientUUID)
+	log.Debug("Deleted engine client proxy")
+}
+
+func (h *EngineClientProxyHandler) getNewClientUUID(clients map[string]EngineClientProxy) (string, error) {
+	if len(clients) >= MaxProxyClientPerImage {
+		return "", errors.Errorf("exceed max engine client proxy per image limit: %v", MaxProxyClientPerImage)
+	}
+
+	canidateID := util.UUID()
+	if _, exist := clients[canidateID]; exist {
+		return "", errors.Errorf("proxy(%v) already exist", canidateID)
+	}
+
+	return canidateID, nil
+}
+
+func (h *EngineClientProxyHandler) getHandledEngineClientProxyPools() map[string]engineClientProxyPool {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	return h.proxyPools
+}
+
+func (h *EngineClientProxyHandler) getHandledEngineClientProxyPoolByImage(name string) (engineClientProxyPool, bool) {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	pool, exist := h.proxyPools[name]
+	return pool, exist
+}
+
+func (h *EngineClientProxyHandler) getEngineClientProxy(instanceManagerImage string, uuid string) (engineClientProxy, bool) {
+	h.lock.RLock()
+	defer h.lock.RUnlock()
+
+	client, exist := h.proxyPools[instanceManagerImage].clients[uuid]
+	return client.getEngineClientProxy(), exist
+}
+
+func (h *EngineClientProxyHandler) NewEngineClientProxy(im *longhorn.InstanceManager) error {
+	imImage := im.Spec.Image
+	log := getLoggerForProxyHandler(h.logger, im)
+
+	_, handlerExist := h.getHandledEngineClientProxyPoolByImage(imImage)
+	if !handlerExist {
+		log.Debug("Initializing engine client proxy handler")
+		h.lock.Lock()
+		h.proxyPools[imImage] = engineClientProxyPool{
+			clients:             make(map[string]EngineClientProxy),
+			instanceManagerName: im.Name,
+		}
+		h.lock.Unlock()
+	}
+
+	pool, _ := h.getHandledEngineClientProxyPoolByImage(imImage)
+	clientUUID, err := h.getNewClientUUID(pool.clients)
+	if err != nil {
+		return err
+	}
+
+	h.lock.Lock()
```

Comment:
Maybe you can just make one lock for this method, because creating a client proxy needs atomic. Granular lock controlling here seems not beneficial a lot.

Reply:
Done.